### PR TITLE
Update order calculations

### PR DESCRIPTION
**Pull request by IKarbowiak**
Creates order line discounted objects for promotions.

Part of https://github.com/saleor/saleor/issues/13316

- Add `OrderLineDiscount`
- Create `OrderLineDiscount` during checkout completion and in `create_order_line` method
- Use the `discounted_price` from variant channel listings for order calculations

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are either absent or guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migrations are either absent or optimized for zero downtime
* [ ] The changes are covered by test cases

### CODE REVIEW

File: saleor/order/utils.py [hunk 1]
```diff
@@ -324,14 +319,39 @@ def create_order_line(
...
+            OrderLineDiscount(
+                line=line,
+                type=DiscountType.PROMOTION,
+                value_type=rule.reward_value_type,
+                value=rule.reward_value,
+                amount_value=rule_discount_amount,
+                currency=line.currency,
+                name=get_discount_name(rule, rule_info.promotion),
+                # TODO: set the promotion translation
+                translated_name="",
```
Comment (IKarbowiak) [inline saleor/order/utils.py:339]: Will be done in separate PR (branch: `set-the-promotion-translation-in-line-discounts`)

**Review by korycins — APPROVED** [review]

**Review by 8r2y5 — APPROVED** [review]

**Review by zedzior — APPROVED** [review]

File: saleor/checkout/tests/test_order_from_checkout.py [hunk 2]
```diff
@@ -723,3 +723,46 @@ def delete_checkout(*args, **kwargs):
...
+        app=app,
+    )
+
+    # then
+    assert order.lines.count() == 1
+    line = order.lines.first()
+    assert line.discounts.count() == 1
+    assert line.sale_id
+    assert line.unit_discount_amount
+    assert line.discounts.count() == 1
```
Comment (fowczarek) [inline saleor/checkout/tests/test_order_from_checkout.py:763]: Should we also check `unit_discount_reason` here?

**Review by SzymJ — APPROVED** [review]

File: saleor/graphql/order/mutations/order_lines_create.py [hunk 3]
```diff
@@ -197,6 +208,24 @@ def perform_mutation(  # type: ignore[override]
 
         return OrderLinesCreate(order=order, order_lines=added_lines)
 
+    @classmethod
+    def get_variant_rule_info_map(cls, data, channel_id, language_code):
```
Comment (fowczarek) [inline saleor/graphql/order/mutations/order_lines_create.py:212]: @IKarbowiak how about adding one benchmark for this function to confirm that we don't perform multiple database queries here?

**Review by fowczarek — APPROVED** [review]

File: saleor/graphql/order/mutations/order_lines_create.py (hunk 3 above)
Comment (IKarbowiak) [reply saleor/graphql/order/mutations/order_lines_create.py:212]: Done
